### ISSUE TITLE
AsOf Join Performance Improvements

### ISSUE THREAD

**Issue by hawkfish** [duckdb/duckdb#7187]
Profiling the benchmark almost all the time seems to be going here - in particular when I modify the benchmark to only join on the time column (`SELECT SUM(v) FROM probe ASOF JOIN build USING(t);`). It seems the random access is hurting performance.

Does it perhaps make sense to fully sort the left-side as well? Or perhaps to batch input from the left-side in larger buffers (e.g. 1M rows instead of 2K rows) and sorting that to reduce random access?

_Originally posted by @Mytherin in https://github.com/duckdb/duckdb/pull/7153#discussion_r1172201109_

**Comment by hawkfish** [comment]
Follow on comments

@hawkfish https://github.com/duckdb/duckdb/pull/7153#discussion_r1172727374
@lnkuiper https://github.com/duckdb/duckdb/pull/7153#discussion_r1173497946

**Comment by tdoehmen** [comment]
Hi, I gave asof joins a try as I'm quite excited about the feature 🙂 !! We are using this kind of joins quite extensively in our feature store (https://www.hopsworks.ai/the-python-centric-feature-store) to join machine learning features in a point-in-time correct manner. 

So I ran some early benchmarks and wanted to share them with you. I used the 0.7.2.dev2675 build which I hope catched your Physical Join operator PR. 

It seems that with a growing left side, the naive approach (join+windows function) still has an edge over the native asof joins. It's quite possible that that's related to the random access of the left side, but not sure. Below are some numbers 

| nrows base table | nrows lookup table | window function | asof   |
|-----------------|------------------|-----------------|--------|
| 10              | 1M               | 0.28s           | 0.18s  |
| 1000            | 1M               | 0.35s           | 0.85s  |
| 10000           | 1M               | 0.37s           | 6.19s  |
| 100000          | 1M               | 0.57s           | 70.95s |

I also attached the code I used for testing, hope it can be of use 🙂 Thanks for your work on this cool feature, looking forward to make good use of it in the future. 

[asof_join_benchmark.zip](https://github.com/duckdb/duckdb/files/11365832/asof_join_benchmark.zip)

**Comment by hawkfish** [comment]
Thanks for the report! 

While initially alarming, after digging into it I don't think it is something that needs fixing in the sort term. Here's my analysis, (and please correct me if I have missed something!)

Simplifying the comparison query to leave out the unused payload columns (the copying cost will be about the same for all three versions):

```sql
CREATE TABLE probe AS
    SELECT r AS id, 
        '2021-01-01T00:00:00'::TIMESTAMP + INTERVAL (random() * 60 * 60 * 24 * 365) SECOND 
            AS probe_ts_1_0,
    FROM range(0, 100000) tbl(r);

CREATE TABLE build AS
    SELECT r % 100000 AS id, 
        '2021-01-01T00:00:00'::TIMESTAMP + INTERVAL (random() * 60 * 60 * 24 * 365) SECOND 
            AS build_ts_1_0,
        (random() * 100000)::INTEGER AS build_int_1_0
    FROM range(0, 1000000) tbl(r);

SELECT
    probe.*,
    build.*,
    rank() OVER (PARTITION BY "probe"."id", "probe"."probe_ts_1_0" ORDER BY "build"."build_ts_1_0" DESC) 
        AS pit_rank
FROM       "probe"
INNER JOIN "build"
ON         "probe"."id" = "build"."id"
AND        "probe"."probe_ts_1_0" >= "build"."build_ts_1_0"
QUALIFY    "pit_rank" = 1
```

(The table names confused me here... `base` is the `probe` table and `lookup` is the `build` table?)

Tossing this into the benchmark harness, I got these results:

| Algorithm | Median of 5 |
| :--- | ---: |
| IEJoin | 22.72 |
| AsOf | 12.29 |
| Inner Join | 0.03 |

This query finds all the `build` values with the same `id` and a smaller timestamp, then ranks them and picks the smallest. It does this by applying a 50% selective predicate to a 10% selective hash join that produces 10 matches per probe, then partitioning and sorting the already well-partitioned values (on `id`) that come out of the HJ. Put into the usual stock price lookup terms, this is looking up a single price for 100K stocks (`id`) that have 10 prices each. 

Is this a common scenario? We should definitely look at it at some point, but it wasn't the use case I was optimizing for on the first pass of the feature. Consider a different stock pricing scenario, where a portfolio has 100 stocks and a year's worth of pricing data down to the minute. The hash join here would simply melt down as the build side would have 120K entries in each bucket. The probe table would be fanned out by 60K times (assuming the inequality has 50% selectivity) after a lot of horribly slow pointer chasing.

So this is a great report but it will need to wait. The current perf is not optimal, but it has no fanout issues (probes produce at most one value) and that seems safest for the next release. But at the very least we can add it to the benchmark suite for later work.

**Comment by hawkfish** [comment]
One funny coda here...I noticed that the HJ was parallel but AJ was single-threaded. This is because the tables can be swapped for HJ (which is symmetric) so it ends up probing with the larger table, which is large enough to trigger parallel scanning. AJ is asymmetric so it can't benefit from that at. this scale.

**Comment by hawkfish** [comment]
In this test case, the swap results in the right hand side having no chains at all (because the `id` column is unique) and it is very fast. But if you were querying for more than one price per id, it could get ugly.

Still it would be nice to be able to rewrite this case if we could detect it, but I'm not sure our stats are up to the task right now.

**Comment by hawkfish** [comment]
Joining the two parquet files from the original Jupyter notebook (|build| == 1,000,000, |probe| == 100,000) with `master` now runs in ~3s on my MacBook Pro:

```sql
CREATE VIEW "base" AS FROM read_parquet('base_table.parquet');
CREATE VIEW "lookup" AS FROM read_parquet('lookup_table.parquet');

WITH right_lookup AS
(SELECT ...
       FROM       "base"
       ASOF  JOIN "lookup"
       ON         "base"."id" = "lookup"."id"
       AND        "base"."base_ts_1_0" >= "lookup"."lookup_ts_1_0" ) 
(
       SELECT ...
       FROM   right_lookup
       ORDER BY 1, 2)
;
```

The notebook says it runs in 70s, so this seems like progress.

**Comment by tdoehmen** [comment]
Looks promising! I'll try it out soon :)

**Comment by hawkfish** [comment]
Looking at a synthetic version of your feature lookup example, I'm not seeing any advantage to the window/inner join now. In fact, the 1000:1 ratio takes forever and has to spool to disk because of the combinatorial explosion. I'm not sure why it didn't need to do that before?

My test data looks like this:

```sql
-- 10K users and ~15 timestamps
CREATE TABLE probe15 AS
	SELECT customer_id, purchase_timestamp
	FROM range(10000) c(customer_id), 
	     range('2022-01-01'::TIMESTAMP, 
                        '2023-01-01'::TIMESTAMP, 
                        INTERVAL 26 DAY) 
               t(purchase_timestamp)

-- 10K users and 1 timestamp
CREATE TABLE probe1 AS
	SELECT customer_id, '2022-01-01'::TIMESTAMP + INTERVAL (customer_id) HOUR purchase_timestamp
	FROM range(10000) c(customer_id)
```

For the build/lookup data, I made three tables that looked like this:

```sql
CREATE TABLE build1000 AS
	SELECT customer_id, feature_generated_timestamp, (RANDOM() * 1000)::DECIMAL(7,2) AS revenue
	FROM range(10000) c(customer_id), 
	     range('2022-01-01'::TIMESTAMP, 
                        '2023-01-01'::TIMESTAMP, 
                        INTERVAL 35 MINUTE)
                 t(feature_generated_timestamp)
-- INTERVAL 350 MINUTE
-- INTERVAL 59 HOUR
```

which give ratios of 10:1, 100:1 and 1000:1 for the 15 element probe and 150:1, 1500:1 and 15000:1 for the single element probe. The results are:

 | Ratio | AsOf | IEJoin | Inner |
 | ----: | -----: | -----: | -----: |
 | 10 | 0.856 | 6.139 | 1.450
 | 100 | 1.644 | 8.245 | 8.944 |
 | 1000 | 10.699 | 44.209 | 365.600 |
| 150 | 0.263 | 0.825 | 0.236 |
| 1500 | 1.070 | 2.857 | 0.736 |
| 15000 | 10.297 | 36.931 | 7.179 |

(Median of 5 except for Inner/1000). 

The inner join plan looks completely reasonable:

```
┌───────────────────────────┐                             
│         PROJECTION        │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│__internal_decompress_integ│                             
│     ral_bigint(#0, 0)     │                             
│             #1            │                             
│             #2            │                             
└─────────────┬─────────────┘                                                          
┌─────────────┴─────────────┐                             
│          ORDER_BY         │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│          ORDERS:          │                             
│    win.customer_id ASC    │                             
│ win.purchase_timestamp ASC│                             
└─────────────┬─────────────┘                                                          
┌─────────────┴─────────────┐                             
│         PROJECTION        │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│__internal_compress_integra│                             
│     l_usmallint(#0, 0)    │                             
│             #1            │                             
│             #2            │                             
└─────────────┬─────────────┘                                                          
┌─────────────┴─────────────┐                             
│         PROJECTION        │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│        customer_id        │                             
│     purchase_timestamp    │                             
│          revenue          │                             
└─────────────┬─────────────┘                                                          
┌─────────────┴─────────────┐                             
│         PROJECTION        │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│             #0            │                             
│             #2            │                             
│             #3            │                             
└─────────────┬─────────────┘                                                          
┌─────────────┴─────────────┐                             
│           FILTER          │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│       (pit_rank = 1)      │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│       EC: 1510561255      │                             
└─────────────┬─────────────┘                                                          
┌─────────────┴─────────────┐                             
│           WINDOW          │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│ RANK() OVER (PARTITION BY │                             
│        customer_id,       │                             
│ purchase_timestamp OR...  │                             
│ feature_generated_tim...  │                             
│            LAST)          │                             
│ RANK() OVER (PARTITION BY │                             
│        customer_id,       │                             
│ purchase_timestamp OR...  │                             
│ feature_generated_tim...  │                             
│            LAST)          │                             
└─────────────┬─────────────┘                                                          
┌─────────────┴─────────────┐                             
│         HASH_JOIN         │                             
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             
│           INNER           │                             
│ customer_id = customer_id │                             
│feature_generated_timestamp├──────────────┐              
│    <= purchase_timestamp  │              │              
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │              │              
│       EC: 1510561255      │              │              
│      Cost: 1510561255     │              │              
└─────────────┬─────────────┘              │                                           
┌─────────────┴─────────────┐┌─────────────┴─────────────┐
│         SEQ_SCAN          ││         SEQ_SCAN          │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│         build1000         ││           probe           │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│        customer_id        ││        customer_id        │
│feature_generated_timestamp││     purchase_timestamp    │
│          revenue          ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││         EC: 150000        │
│       EC: 150180000       ││                           │
└───────────────────────────┘└───────────────────────────┘                             

```

So to sum up the current state of affairs:

* For all ratios with 15 probes, AsOf is the most performant.
* For small ratios with 15 probes, Inner join beats IEJoin (both with windowing), but by 100:1 it is starting to explode. 
* For single element probes, Inner join is most effective, but even there, its edge over AsOf is only about 50% at scale.

So it seems that AsOf could be improved upon, but predicting where that happens could be tricky, with incredibly bad down sides.

**Comment by tdoehmen** [comment]
I see what you mean. I think what is different in your dataset is that the key you are doing the equality join on is not unique on the probe side. In my initial example dataset it is unique. 

I actually stubled over a similar dataset recently. It is quite small, with 1M entries on the probe side, and ~30k on the build side. The key for the equality join has only 259 distinct values, though. The query quickly blows up memory to >16GB and then fails at some point with OOM, even with disk-offloading switched on. 

The ASOF join works fine, finishes in under a second.

```
CREATE OR REPLACE TABLE probe AS
	SELECT * FROM read_parquet("raw_1.parquet")

CREATE OR REPLACE TABLE build AS
	SELECT * FROM read_parquet("dropoff_1.parquet")
```

```
SELECT
    probe.*,
    build.*,
    rank() OVER (PARTITION BY "probe"."do_location_id", "probe"."tpep_dropoff_datetime" ORDER BY "build"."ts" DESC) 
        AS pit_rank
FROM       "probe"
INNER JOIN "build"
ON         "probe"."do_location_id" = "build"."location_id"
AND        "probe"."tpep_dropoff_datetime" >= "build"."ts"
QUALIFY    "pit_rank" = 1
```

[pit-join-example.zip](https://github.com/duckdb/duckdb/files/12152869/pit-join-example.zip)

**Comment by hawkfish** [comment]
I think this issue has been beaten to death now. There will always be room for improvement but I think we should move on to epic failures.
